kernel/src/threading/interrupt.rs

```diff
@@ -0,0 +1,461 @@
+    pic.intr_names[19] = "#XF SIMD Floating-Point Exception";
+
+    pic
+}
+
+impl Pic {
+    /// Returns true if interrupts are enabled, false otherwise.
+    pub fn get_intr_enabled(&self) -> bool {
+        let flags: u32;
+
+        // Push the flags register on the processor stack, then pop the value off the stack into
+        // `flags'.
+        //
+        // See
+        // [Intel® 64 and IA-32 Architectures Software Developer's Manual Volume 2B: Instruction Set Reference, M-U](https://www.intel.com/content/www/us/en/content-details/825757/)
+        // "PUSHF/PUSHFD/PUSHFQ—Push EFLAGS Register Onto the Stack" and "POP—Pop a Value From the Stack" and
+        // [Intel® 64 and IA-32 Architectures Software Developer's Manual Volume 3A: System Programming Guide, Part 1](https://www.intel.com/content/www/us/en/content-details/825758/)
+        // 6.8.1 "Masking Maskable Hardware Interrupts"
+        unsafe {
+            asm!("pushf; pop {0}", out(reg) flags);
+        }
+
+        flags & 0x200 != 0
+    }
+
+    /// Enables or disables interrupts as specified by `enable'.
+    ///
+    /// Returns the previous interrupt state.
+    pub fn set_intr_enabled(&self, level: bool) -> bool {
+        if level {
+            self.iter_enable()
+        } else {
+            self.intr_disable()
+        }
+    }
+
+    /// Enables interrupts and returns the previous interrupt state.
+    pub fn iter_enable(&self) -> bool {
+        let old_level = self.get_intr_enabled();
+        assert!(!self.intr_context());
+
+        unsafe {
+            asm!("sti");
+        }
+        old_level
+    }
+
+    /// Disables interrupts and returns the previous interrupt status
+    pub fn intr_disable(&self) -> bool {
+        let old_level = self.get_intr_enabled();
+
+        // Disable interrupts by clearing the interrupt flag.
+        //
+        // See
+        // [Intel® 64 and IA-32 Architectures Software Developer's Manual Volume 2A: Instruction Set Reference, A-L](https://www.intel.com/content/www/us/en/content-details/825747/)
+        // "CLI—Clear Interrupt Flag" and
+        // [Intel® 64 and IA-32 Architectures Software Developer's Manual Volume 3A: System Programming Guide, Part 1](https://www.intel.com/content/www/us/en/content-details/825758/)
+        // 6.8.1 "Masking Maskable Hardware Interrupts".
+        unsafe {
+            asm!("cli", options(nostack, preserves_flags));
+        }
+
+        old_level
+    }
+    fn register_handler(
+        &mut self,
+        vec_no: u8,
+        dpl: u8,
+        level: bool,
+        handler: IntrHandlerFunc,
+        name: &'static str,
+    ) {
+        if level {
+            unsafe {
+                self.idt[vec_no as usize] = self.make_trap_gate(INTR_STUBS[vec_no as usize], dpl);
+            }
+        } else {
+            unsafe {
+                self.idt[vec_no as usize] = self.make_intr_gate(INTR_STUBS[vec_no as usize], dpl);
+            }
+        }
```

Comment:
Nit: Maybe rename `level` to `enabled` if we are going to make it a `bool`? Optionally, maybe get rid of some of this repetition if you feel comfortable with this. 

```suggestion
        let stub = INTR_STUBS[vec_no as usize];
        
        unsafe {
            self.idt[vec_no as usize] = if level {
                self.make_trap_gate(stub, dpl);
            } else {
                self.make_intr_gate(stub, dpl);
            };
        }
```

Reply:
Yes sorry I forgot to rename `level` to `enabled`. Will change that in a later update.